api/v1_queues.inc

```diff
@@ -0,0 +1,125 @@
+<?php
+include_once($relPath.'Project.inc');
+include_once($relPath.'release_queue.inc');
+include_once($relPath.'user_is.inc');
+include_once("exceptions.inc");
+include_once("misc.inc");
+
+// DP API v1 -- Queues
+
+//---------------------------------------------------------------------------
+// queues -- list queues, optionally filtered by round
+function api_v1_queues($method, $data, $query_params)
+{
+    global $Round_for_round_id_;
+
+    $roundid = array_get($query_params, "roundid", null);
+    if (!is_null($roundid)) {
+        try {
+            if (!array_key_exists($roundid, $Round_for_round_id_)) {
+                throw new InvalidRoundException("Invalid round");
+            }
+            $round = $Round_for_round_id_[$roundid];
+        } catch (InvalidRoundException $exception) {
+            throw new NotFoundError($exception->getMessage(), $exception->getCode());
+        }
+    } else {
+        $round = null;
+    }
+
+    $show = get_enumerated_param($query_params, "show", "all", ["enabled", "populated", "all"]);
+
+    // Fields we  can query on
+    $valid_fields = [
+        "name",
+        "group",
+    ];
+
+    $valid_return_fields = [
+        "id",
+        "round_id",
+        "group",
+        "ordering",
+        "name",
+        "enabled",
+        "populated",
+        "length",
+        "unheld_length",
+        "comment",
+    ];
+
+    if (user_can_see_queue_settings()) {
+        array_push($valid_return_fields, "project_selector", "release_criterion");
+    }
+
+    // restrict to list of desired fields, if set
+    $return_fields = array_get_as_array($query_params, "field", $valid_return_fields);
+    $return_fields = array_intersect($return_fields, $valid_return_fields);
+
+    $name_selector = array_get($query_params, "name", null);
+    $group_selector = array_get($query_params, "group", null);
+
+    $output = [];
+    foreach (fetch_queues_data($round, $show, false, $name_selector, $group_selector) as $queue_data) {
+        array_remove_invalid_fields($queue_data, $return_fields);
+        $output[] = $queue_data;
+    }
+    return $output;
+}
+
+//---------------------------------------------------------------------------
+// queues/:queueid -- return release queue info
+function api_v1_queue($method, $data, $query_params)
+{
+    $queue = $data[":queueid"];
+
+    return $queue;
+}
+
+//---------------------------------------------------------------------------
+// queue/:queueid/stats -- return release queue stats
+function api_v1_queue_stats($method, $data, $query_params)
+{
+    global $Round_for_round_id_;
+    $queue = $data[":queueid"];
+
+    $round = $Round_for_round_id_[$queue["round_id"]];
```

Comment:
You could probably use `validate_round()` here too.

Reply:
Done